File: frontend/lib/validators/academic-year.ts
import { z } from "zod";

import { trimmedString } from "@/lib/validators/shared";

export const academicYearFormSchema = z
  .object({
    label: trimmedString()
      .length(9, "Use a format like 2025-2026")
      .regex(/^\d{4}-\d{4}$/, "Use a format like 2025-2026"),
    is_active: z.boolean(),
  })
  .refine(
    (value) => {
      const [startYear, endYear] = value.label.split("-").map(Number);
      return endYear === startYear + 1;
    },
    {
      message: "Academic year must span exactly one year",
      path: ["label"],
    }
  );

export type AcademicYearFormValues = z.infer<typeof academicYearFormSchema>;

File: backend/app/schemas/academic_year.py
import uuid
from datetime import datetime

from pydantic import ConfigDict, Field, field_validator

from app.schemas.common import NonEmptyUpdateModel, StrippedModel


def _is_consecutive_label(value: str) -> bool:
    start_year, end_year = value.split("-")
    return int(end_year) == int(start_year) + 1


class AcademicYearBase(StrippedModel):
    label: str = Field(min_length=9, max_length=9, pattern=r"^\d{4}-\d{4}$", description='e.g. "2025-2026"')
    is_active: bool = True

    @field_validator("label")
    @classmethod
    def validate_consecutive_label(cls, value: str) -> str:
        if not _is_consecutive_label(value):
            raise ValueError("Academic year label must span exactly one year, e.g. 2025-2026")
        return value


class AcademicYearCreate(AcademicYearBase):
    pass


class AcademicYearUpdate(NonEmptyUpdateModel):
    label: str | None = Field(default=None, min_length=9, max_length=9, pattern=r"^\d{4}-\d{4}$")
    is_active: bool | None = None

    @field_validator("label")
    @classmethod
    def validate_consecutive_label(cls, value: str | None) -> str | None:
        if value is not None and not _is_consecutive_label(value):
            raise ValueError("Academic year label must span exactly one year, e.g. 2025-2026")
        return value


class AcademicYearRead(AcademicYearBase):
    model_config = ConfigDict(from_attributes=True)

    id: uuid.UUID
    created_at: datetime
    updated_at: datetime
